File: 2658_maximum_number_of_fish_in_a_grid.java
class Solution {
    public int findMaxFish(int[][] grid) {
        int maxFish = 0, rows = grid.length, cols = grid[0].length;
        for (int i = 0; i < rows; i++)
            for (int j = 0; j < cols; j++)
                if (grid[i][j] > 0)
                    maxFish = Math.max(maxFish, bfs(grid, i, j, rows, cols));
        return maxFish;
    }

    private int bfs(int[][] g, int x, int y, int r, int c) {
        int fishCount = 0;
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[] {x, y});
        while (!q.isEmpty()) {
            int[] cell = q.poll();
            int i = cell[0], j = cell[1];
            if (g[i][j] > 0) {
                fishCount += g[i][j];
                g[i][j] = 0;
                if (i > 0) q.add(new int[] {i - 1, j});
                if (i + 1 < r) q.add(new int[] {i + 1, j});
                if (j > 0) q.add(new int[] {i, j - 1});
                if (j + 1 < c) q.add(new int[] {i, j + 1});
            }
        }
        return fishCount;
    }
}

File: 2658_maximum_number_of_fish_in_a_grid.cpp
class Solution {
public:
    int findMaxFish(vector<vector<int>>& a) {
        int p = 0, v = a.size(), h = a[0].size();
        for (int u = 0; u < v; ++u)
            for (int o = 0; o < h; ++o)
                if (a[u][o]) {
                    int e = 0;
                    queue<pair<int, int>> z;
                    z.push({u, o});
                    while (!z.empty()) {
                        auto [f, g] = z.front();
                        z.pop();
                        if (a[f][g]) {
                            e += a[f][g];
                            a[f][g] = 0;
                            if (f && a[f - 1][g]) z.push({f - 1, g});
                            if (f + 1 < v && a[f + 1][g]) z.push({f + 1, g});
                            if (g && a[f][g - 1]) z.push({f, g - 1});
                            if (g + 1 < h && a[f][g + 1]) z.push({f, g + 1});
                        }
                    }
                    p = max(p, e);
                }
        return p;
    }
};
